### ISSUE TITLE
Users can analyze data from the DCP

### ISSUE THREAD

**Issue by kozbo** [HumanCellAtlas/dcp#337]
**Overview**: Users can obtain the analysis results of existing assays in the DCP and use the data and metadata to perform analysis outside of the DCP

**User Story**: As a researcher with a keyboard I want exemplar code for downloading and comparing data from the Human Cell Atlas with my own data. As a researcher with a keyboard I want to be able to take the search results from the DCP and send references to the files to a workspace in a portal where I can analyze the data with an arbitrary WDL or Jupyter notebook.

**Deliverable at the end of Q2**: 1) Consumer vignettes which support researchers downloading and conducting data comparisons on their own laptop. 2) A "send to Portal" function in the Data Browser. This deliverable also includes 3) documentation on the Data Portal for how to do this.

**Evaluation Plan - how this will be evaluated at the end of the quarter**: 1) There is a tutorial (on or pointed to from our “[Guides](https://dev.data.humancellatlas.org/guides)” portal section) that a researcher can follow that will show them how to search and download data and perform some basic analysis.  A great candidate for this is leveraging the “[Download Any BAM File](https://github.com/HumanCellAtlas/data-consumer-vignettes/tree/master/tasks/Download%20BAM)” [vignette](https://github.com/HumanCellAtlas/data-consumer-vignettes) in a tutorial on our Data Portal site and including a very basic analysis example given some BAMs downloaded from HCA.  2) There is a tutorial (on or pointed to from our “Guides” portal section) that a researcher can follow that will show them how to search and send results (could go with BAM files) to Terra.  The tutorial will then show them how to do some basic analysis on the BAMs in Terra via a WDL workflow (could be the same super simple analysis that the first tutorial did).  3) STRETCH There is a tutorial (on or pointed to from our “Guides” portal section) that a researcher can follow that will show them how to search and generate a URL for a matrix of the expression data for their search results.  They can then take this to a Jupyter notebook service running in Terra and execute a vignette based on what we did in the “[HCA Scanpy Demo](https://github.com/HumanCellAtlas/data-consumer-vignettes/tree/master/nov2018_demo)” from Nov 2018.

**We need to consider:**
~* in Q3 putting these tutorials/vignettes under continuous integration testing~ Created [Need CI/CD system for vignettes](https://github.com/HumanCellAtlas/data-consumer-vignettes/issues/28)
* we need to make sure there's a link between the manifest from the data explorer and the metadata TSV for a project.  Need to be able to link these for these tutorials to work
* when a user downloads a manifest, a matrix, and the metadata for a given project we need to make sure a researcher can join between all three files successfully

**Needs from Responsible Groups**: 

~* TJ - Query Service~
  ~* TBD: we need to better flesh this out with TJ, I don’t think we have any requirements per se but more of a desire to see how we might leverage the query service~ 
  ~* when a user downloads a manifest, a matrix, and the metadata for a given project we need to make sure a researcher can join between all three files successfully~
  ~* Help write/validate tutorials~
~* Ambrose - Matrix Service~
  ~* We need the matrix service to work with the queries specified in the tutorials.~  
  ~* Ideally the matrix service would be fast overall (<1 min to render) and including caching to further speed results that have previously been requested.~
  ~* Help write/validate tutorials~
  ~* when a user downloads a manifest, a matrix, and the metadata for a given project we need to make sure a researcher can join between all three files successfully~
* Brian - Data Browser/Data Store
  * We need handoff to Terra so search results in the portal can create workspaces in the Terra platform, available for arbitrary analysis using WDL (tutorial 2 above)
  * We need something from tutorial 1 that we can write a WDL for, UCSC can write the WDL version given our experience.  The Broad volunteered to help with this as well.
    * We need a URL for a matrix service produced so we can cut and paste into a Jupyter notebook (tutorial 3 above) -- DONE, I see URLs are created as part of requesting a matrix such as [this one](https://s3.amazonaws.com/dcp-matrix-service-results-staging/f37a15f5f3d7125bad59340fc0626364a72cfc05be3cf876a375ba6f82fa325e.loom).  Not sure how long the URLs last for.
      * handle the Jupyter notebook server setup via Terra
      * Updated vignette for the Jupyter notebook analysis
      * The UI on the Browser might need to be tweaked to better support slow matrix responses if the speed improvements don’t get us below 1 min
    * We need to write the tutorials with the partners above
  * when a user downloads a manifest, a matrix, and the metadata for a given project we need to make sure a researcher can join between all three files successfully
* Unassigned:
  ~* We need someone to take point on creating (updating) a vignette for use in Terra’s Jupyter notebook system, really like this to be an updated version of Genevieve's notebook from our previous demo (“HCA Scanpy Demo”)... from our conversation on 20190425 none of the PLs/POs volunteered for this so this is now a stretch goal since we don't have a clear group to own this.~

**Note**: I confirmed this proposal above with @kozbo and @theathorn on 20190417 and they have/will link appropriate Epics from across their teams (and ask other teams to do the same) to these theme epic via ZenHub to track the work to be done in Q2.

**For more information see**: https://docs.google.com/spreadsheets/d/1iAL2JR3ndgMmYwojUU0kU8Pm7B7ycRLt_973seREd3M

**Comment by briandoconnor** [comment]
Note, this Epic was created as part of our Q2 and beyond roadmapping process.  It has some overlap with other Epics and Kevin will note/link in Zenhub.  Also, there may be individual stories created and linked to this epic in various teams identified in the roadmap planning process.

**Comment by theathorn** [comment]
From https://docs.google.com/spreadsheets/d/1iAL2JR3ndgMmYwojUU0kU8Pm7B7ycRLt_973seREd3M Label G Deliverable: "Consumer vignettes which support researchers downloading and conducting data comparisons on their own laptop, A "send to Portal" button in the Data Browser.  This deliverable also includes documentation on the portal for how to do this".

**Comment by tjchen** [comment]
For my item:

Given the sheer amount of work in Q2 by everyone and the limited window, the Query Service Team's proposal is to spend most of the quarter on (1) UX interviews to ensure that MySQL is easier to use than ES, and (2) ensure we have clear search use cases and beta test with these teams.

In future quarters I see us integrating Query Service into this flow, but this Epic is not dependent on Query Service for v1. We can help with search use cases, and I know Genevieve has already put a substantial amount of effort into this.

FYI @mweiden  @GenevieveHaliburton @kislyuk @MDunitz

**Comment by tburdett** [comment]
I have two questions:
1) What constitutes "data" here?
2) How is metadata made available to users outside of the DCP?

Answers from Brian O:
1) In 1. in the evaluation plan, this applies to all data files - so do a search, get a file manifest, and export these files to <environment>. In 3. in the evaluation plan, this applies very specifically to the matrix service outputs
2) There is a key requirement that there is some link between the file manifest produced by the browser and the metadata provided to consumers e.g. as part of the TSV

**Comment by briandoconnor** [comment]
We had a chat today with Ambrose, Adrienne, David R. Calvin, Marcus, Kevin, and Trevor (sorry if I missed anyone).  Points from this meeting:

* we're on track for delivering all these features except (2) the send to Terra feature
* the delay is caused by slow checkout in google but this is being mitigated now and we have a path forward
* the feature will likely be available at the beginning of Q3

**Comment by brianraymor** [comment]
Updating to _Blocked_ for visibility. See [comment](https://github.com/HumanCellAtlas/matrix-service/issues/291#issuecomment-500003159).

**Comment by theathorn** [comment]
The "slow checkout in Google' is resolved but we are essentially blocked by the very slow response from the Terra team in resolving issues on their side. I am working closely with Cara on this but we don't have an ETA to resolve.

**Comment by theathorn** [comment]
Unblocking since we have an end-to-end Terra prototype working.

**Comment by theathorn** [comment]
4 remaining issues on track to complete in M3.

**Comment by theathorn** [comment]
Noting that the "Export to Terra" button is now "live" in the Data Browser.

**Comment by theathorn** [comment]
Just need the Terra tutorial to be reviewed and merged.
